### PR TITLE
Adding OpenUI5 example

### PR DESCRIPTION
**Pull request by agraebe**
### About

OpenUI5 is an Open Source JavaScript UI library, maintained by SAP. OpenUI5 lets you build enterprise-ready web applications, responsive to all devices, running on almost any browser of your choice. It’s based on JavaScript, using JQuery as its foundation and follows web standards. It eases your development with a client-side HTML5 rendering library including a rich set of controls and supports data binding to different models (JSON, XML and OData).
### Previous PRs

[I wanted to update](https://github.com/tastejs/todomvc/pull/1413) the unmaintained and outdated submission for SAPUI5 and realized it needs a rework from the ground up for several reasons:
- existing SAPUI5 example is [not alive and still in "Labs" mode](https://github.com/tastejs/todomvc/issues/1110)
- exisiting SAPUI5 example is based on outdated app template
- existing SAPUI5 example is based on outdated OpenUI5 and [doesn't work](http://todomvc.com/examples/sapui5/)
- existing SAPUI5 example [doesn't pass the automated browser tests](https://github.com/tastejs/todomvc/pull/1327)
- existing SAPUI5 example lacks a good documentation
- the opensourced version of SAPUI5 is called [OpenUI5](http://openui5.org/faq.html)
### Why should this PR be accepted?

The example for SAPUI5 needs to be replaced with this OpenUI5 example for the several reasons mentioned above. Furthermore, this example app is implemented by closely following the most current best practices (app descriptor, router, i18n, component-based, localStorage wrapper, custom control). The implementation follows the contribution guidelines, the suggested codestyle, and it was  JSHint, JSCS, JSONLint, CSSLint, XMLLint checked and W3C validated.
### Live Demo

[Here's a live demo hosted on Plunker](http://embed.plnkr.co/yK4QztSonwYrHQ5cYtD1/).
### Tests

All the current tests pass on latest Firefox & Chrome. Here's the output for `npm run test -- --framework=openui5`

```
> mocha allTests.js --no-timeouts --reporter spec "--framework=openui5"



  TodoMVC - openui5
    When page is initially opened
      ✓ should focus on the todo input field
    No Todos
      ✓ should hide #main and #footer (94ms)
    New Todo
      ✓ should allow me to add todo items (732ms)
      ✓ should clear text input field when an item is added (301ms)
      ✓ should append new items to the bottom of the list (927ms)
      ✓ should trim text input (333ms)
      ✓ should show #main and #footer when items added (296ms)
    Mark all as completed
      ✓ should allow me to mark all items as completed (149ms)
      ✓ should correctly update the complete all checked state (233ms)
      ✓ should allow me to clear the completion state of all items (192ms)
      ✓ complete all checkbox should update state when items are completed / cleared (261ms)
    Item
      ✓ should allow me to mark items as complete (685ms)
      ✓ should allow me to un-mark items as complete (683ms)
    Editing
      ✓ should focus the input
      ✓ should hide other controls when editing
      ✓ should save edits on enter (473ms)
      ✓ should save edits on blur (542ms)
      ✓ should trim entered text (586ms)
      ✓ should remove the item if an empty text string was entered (415ms)
      ✓ should cancel edits on escape (461ms)
    Counter
      ✓ should display the current number of todo items (548ms)
    Clear completed button
      ✓ should display the correct text (104ms)
      ✓ should remove completed items when clicked (256ms)
      ✓ should be hidden when there are no items that are completed (233ms)
    Persistence
      ✓ should persist its data (1881ms)
    Routing
      ✓ should allow me to display active items (341ms)
      ✓ should respect the back button (720ms)
      ✓ should allow me to display completed items (256ms)
      ✓ should allow me to display all items (507ms)
      ✓ should highlight the currently applied filter (278ms)

  30 passing (2m)
```

I excluded the following, repeated console warning (1) and info (6) occurances, coming from `sap-ui-core.js`. I think it's safe to ignore them.

```
[ Object {
    level: Object { name: 'WARNING', value: 900 },
    message: '{"message":{"column":1120,"executionContextId":1,"level":"info","line":16,"parameters":[{"type":"string","value":"2015-12-30 18:39:51 Device API logging initialized - DEVICE"}],"source":"console-api","stackTrace":[{"columnNumber":1120,"functionName":"log","lineNumber":16,"scriptId":"45","url":"http://localhost:8000/examples/openui5/resources/sap-ui-core.js"},{"columnNumber":1298,"functionName":"","lineNumber":16,"scriptId":"45","url":"http://localhost:8000/examples/openui5/resources/sap-ui-core.js"},{"columnNumber":11585,"functionName":"","lineNumber":32,"scriptId":"45","url":"http://localhost:8000/examples/openui5/resources/sap-ui-core.js"}],"text":"2015-12-30 18:39:51 Device API logging initialized - DEVICE","timestamp":1451529591.09247,"type":"log","url":"http://localhost:8000/examples/openui5/resources/sap-ui-core.js"}}',
    timestamp: 1451529591123,
    type: '',
    closure_uid_687222711: undefined },
  Object {
    level: Object { name: 'WARNING', value: 900 },
    message: '{"message":{"column":8411,"executionContextId":1,"level":"info","line":88,"parameters":[{"type":"string","value":"2015-12-30 18:39:51.263060 registerResourcePath (\'\', \'resources/\') -  sap.ui.ModuleSystem"}],"source":"console-api","stackTrace":[{"columnNumber":8411,"functionName":"A","lineNumber":88,"scriptId":"45","url":"http://localhost:8000/examples/openui5/resources/sap-ui-core.js"},{"columnNumber":8797,"functionName":"info","lineNumber":88,"scriptId":"45","url":"http://localhost:8000/examples/openui5/resources/sap-ui-core.js"},{"columnNumber":28044,"functionName":"registerResourcePath","lineNumber":88,"scriptId":"45","url":"http://localhost:8000/examples/openui5/resources/sap-ui-core.js"},{"columnNumber":27433,"functionName":"registerModulePath","lineNumber":88,"scriptId":"45","url":"http://localhost:8000/examples/openui5/resources/sap-ui-core.js"},{"columnNumber":3249,"functionName":"Q.extend.each","lineNumber":60,"scriptId":"45","url":"http://localhost:8000/examples/openui5/resources/sap-ui-core.js"},{"columnNumber":28393,"functionName":"","lineNumber":88,"scriptId":"45","url":"http://localhost:8000/examples/openui5/resources/sap-ui-core.js"},{"columnNumber":33446,"functionName":"","lineNumber":88,"scriptId":"45","url":"http://localhost:8000/examples/openui5/resources/sap-ui-core.js"},{"columnNumber":43787,"functionName":"","lineNumber":88,"scriptId":"45","url":"http://localhost:8000/examples/openui5/resources/sap-ui-core.js"}],"text":"2015-12-30 18:39:51.263060 registerResourcePath (\'\', \'resources/\') -  sap.ui.ModuleSystem","timestamp":1451529591.26416,"type":"log","url":"http://localhost:8000/examples/openui5/resources/sap-ui-core.js"}}',
    timestamp: 1451529591265,
    type: '',
    closure_uid_687222711: undefined },
  Object {
    level: Object { name: 'WARNING', value: 900 },
    message: '{"message":{"column":8411,"executionContextId":1,"level":"info","line":88,"parameters":[{"type":"string","value":"2015-12-30 18:39:51.264449 registerResourcePath (\'ToDoMVC\', \'js/\') -  sap.ui.ModuleSystem"}],"source":"console-api","stackTrace":[{"columnNumber":8411,"functionName":"A","lineNumber":88,"scriptId":"45","url":"http://localhost:8000/examples/openui5/resources/sap-ui-core.js"},{"columnNumber":8797,"functionName":"info","lineNumber":88,"scriptId":"45","url":"http://localhost:8000/examples/openui5/resources/sap-ui-core.js"},{"columnNumber":28044,"functionName":"registerResourcePath","lineNumber":88,"scriptId":"45","url":"http://localhost:8000/examples/openui5/resources/sap-ui-core.js"},{"columnNumber":27433,"functionName":"registerModulePath","lineNumber":88,"scriptId":"45","url":"http://localhost:8000/examples/openui5/resources/sap-ui-core.js"},{"columnNumber":3249,"functionName":"Q.extend.each","lineNumber":60,"scriptId":"45","url":"http://localhost:8000/examples/openui5/resources/sap-ui-core.js"},{"columnNumber":28393,"functionName":"","lineNumber":88,"scriptId":"45","url":"http://localhost:8000/examples/openui5/resources/sap-ui-core.js"},{"columnNumber":33446,"functionName":"","lineNumber":88,"scriptId":"45","url":"http://localhost:8000/examples/openui5/resources/sap-ui-core.js"},{"columnNumber":43787,"functionName":"","lineNumber":88,"scriptId":"45","url":"http://localhost:8000/examples/openui5/resources/sap-ui-core.js"}],"text":"2015-12-30 18:39:51.264449 registerResourcePath (\'ToDoMVC\', \'js/\') -  sap.ui.ModuleSystem","timestamp":1451529591.26563,"type":"log","url":"http://localhost:8000/examples/openui5/resources/sap-ui-core.js"}}',
    timestamp: 1451529591266,
    type: '',
    closure_uid_687222711: undefined },
  Object {
    level: Object { name: 'WARNING', value: 900 },
    message: '{"message":{"column":8411,"executionContextId":1,"level":"info","line":88,"parameters":[{"type":"string","value":"2015-12-30 18:39:51.265379 URL prefixes set to: -  sap.ui.ModuleSystem"}],"source":"console-api","stackTrace":[{"columnNumber":8411,"functionName":"A","lineNumber":88,"scriptId":"45","url":"http://localhost:8000/examples/openui5/resources/sap-ui-core.js"},{"columnNumber":8797,"functionName":"info","lineNumber":88,"scriptId":"45","url":"http://localhost:8000/examples/openui5/resources/sap-ui-core.js"},{"columnNumber":28448,"functionName":"","lineNumber":88,"scriptId":"45","url":"http://localhost:8000/examples/openui5/resources/sap-ui-core.js"},{"columnNumber":33446,"functionName":"","lineNumber":88,"scriptId":"45","url":"http://localhost:8000/examples/openui5/resources/sap-ui-core.js"},{"columnNumber":43787,"functionName":"","lineNumber":88,"scriptId":"45","url":"http://localhost:8000/examples/openui5/resources/sap-ui-core.js"}],"text":"2015-12-30 18:39:51.265379 URL prefixes set to: -  sap.ui.ModuleSystem","timestamp":1451529591.26614,"type":"log","url":"http://localhost:8000/examples/openui5/resources/sap-ui-core.js"}}',
    timestamp: 1451529591266,
    type: '',
    closure_uid_687222711: undefined },
  Object {
    level: Object { name: 'WARNING', value: 900 },
    message: '{"message":{"column":8411,"executionContextId":1,"level":"info","line":88,"parameters":[{"type":"string","value":"2015-12-30 18:39:51.265985   (default) : resources/ -  sap.ui.ModuleSystem"}],"source":"console-api","stackTrace":[{"columnNumber":8411,"functionName":"A","lineNumber":88,"scriptId":"45","url":"http://localhost:8000/examples/openui5/resources/sap-ui-core.js"},{"columnNumber":8797,"functionName":"info","lineNumber":88,"scriptId":"45","url":"http://localhost:8000/examples/openui5/resources/sap-ui-core.js"},{"columnNumber":28495,"functionName":"","lineNumber":88,"scriptId":"45","url":"http://localhost:8000/examples/openui5/resources/sap-ui-core.js"},{"columnNumber":33446,"functionName":"","lineNumber":88,"scriptId":"45","url":"http://localhost:8000/examples/openui5/resources/sap-ui-core.js"},{"columnNumber":43787,"functionName":"","lineNumber":88,"scriptId":"45","url":"http://localhost:8000/examples/openui5/resources/sap-ui-core.js"}],"text":"2015-12-30 18:39:51.265985   (default) : resources/ -  sap.ui.ModuleSystem","timestamp":1451529591.26676,"type":"log","url":"http://localhost:8000/examples/openui5/resources/sap-ui-core.js"}}',
    timestamp: 1451529591267,
    type: '',
    closure_uid_687222711: undefined },
  Object {
    level: Object { name: 'WARNING', value: 900 },
    message: '{"message":{"column":8411,"executionContextId":1,"level":"info","line":88,"parameters":[{"type":"string","value":"2015-12-30 18:39:51.266364   \'ToDoMVC\' : js/ -  sap.ui.ModuleSystem"}],"source":"console-api","stackTrace":[{"columnNumber":8411,"functionName":"A","lineNumber":88,"scriptId":"45","url":"http://localhost:8000/examples/openui5/resources/sap-ui-core.js"},{"columnNumber":8797,"functionName":"info","lineNumber":88,"scriptId":"45","url":"http://localhost:8000/examples/openui5/resources/sap-ui-core.js"},{"columnNumber":28495,"functionName":"","lineNumber":88,"scriptId":"45","url":"http://localhost:8000/examples/openui5/resources/sap-ui-core.js"},{"columnNumber":33446,"functionName":"","lineNumber":88,"scriptId":"45","url":"http://localhost:8000/examples/openui5/resources/sap-ui-core.js"},{"columnNumber":43787,"functionName":"","lineNumber":88,"scriptId":"45","url":"http://localhost:8000/examples/openui5/resources/sap-ui-core.js"}],"text":"2015-12-30 18:39:51.266364   \'ToDoMVC\' : js/ -  sap.ui.ModuleSystem","timestamp":1451529591.26712,"type":"log","url":"http://localhost:8000/examples/openui5/resources/sap-ui-core.js"}}',
    timestamp: 1451529591267,
    type: '',
    closure_uid_687222711: undefined },
  Object {
    level: Object { name: 'WARNING', value: 900 },
    message: 'http://localhost:8000/examples/openui5/resources/sap-ui-core.js 71:91306 Synchronous XMLHttpRequest on the main thread is deprecated because of its detrimental effects to the end user\'s experience. For more help, check http://xhr.spec.whatwg.org/.',
    timestamp: 1451529591363,
    type: '',
    closure_uid_687222711: undefined } ]
```

### CODE REVIEW

**Comment by agraebe** [comment]
In case this PR will get accepted ultimately, I'd follow up with a PR to remove the existing SAPUI5 example app.

**Comment by agraebe** [comment]
The Travis tests exit with:
 ```FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - process out of memory
Aborted (core dumped)```

.. Did you see this before? Any chance to identify the problem for reproduction purposes?

Thanks very much!

**Comment by agraebe** [comment]
We're getting there .. 
```
29 passing (2m)
  1 failing
  1) TodoMVC - openui5 Persistence should persist its data:
     Error: Wait timed out after 5005ms
```

I'm doing some tests with network throttling now to understand what's going wrong here.

**Comment by agraebe** [comment]
Is there anyone who could help sort this out? :) @samccone maybe?

**Comment by samccone** [comment]
I will take a look :)

**Comment by samccone** [comment]
@agraebe looks like you will want to add examples/openui5/js/Component-preload.js to our jscs ignore :)

**Comment by agraebe** [comment]
@samccone Thanks for jumping in ;)

Done. How about this last test that was failing? 

```
  1) TodoMVC - openui5 Persistence should persist its data:
     Error: Wait timed out after 5005ms
```

**Comment by samccone** [comment]
Looks like you need to vendor a few more files..

http://localhost:8000/examples/openui5/node_modules/todomvc-common/base.css
http://localhost:8000/examples/openui5/node_modules/todomvc-app-css/index.css
http://localhost:8000/examples/openui5/node_modules/todomvc-common/base.js

**Comment by agraebe** [comment]
oh boy, you're right.. my gitigmore requires an update! should have checked the logs with more patience :walking:

File: .jscsrc [hunk 1]
```diff
@@ -38,7 +38,8 @@
 		"examples/thorax/js/lib/backbone-localstorage.js",
 		"examples/thorax_lumbar/public/*.js",
 		"examples/typescript-*/js/**/*.js",
-		"examples/vanilladart/**/*.js"
+		"examples/vanilladart/**/*.js",
+		"examples/openui5/js/Component-preload.js"
```
Comment (samccone) [inline .jscsrc]: alphabetical please :)

File: examples/openui5/css/app.css [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+	display: none;
+}
+.sap-desktop .sapMShellCentralBox {
+	box-shadow: none;
+}
+.sapMShellGlobalOuterBackground, .sapMShellGlobalInnerBackground, .sapUshellShellBackground, .sapUiGlobalBackgroundColor, .sapMGlobalBackgroundColor {
+	background-color: transparent;
+	background-image: none;
+}
+.header>h1 {
```
Comment (samccone) [inline examples/openui5/css/app.css]: space between > plz :)

File: examples/openui5/index.html [hunk 3]
```diff
@@ -0,0 +1,20 @@
+<!doctype html>
+<html lang="en" data-framework="openui5">
+
+<head>
+	<meta charset="utf-8">
+	<title>OpenUI5 • TodoMVC</title>
+	<script id="sap-ui-bootstrap" src="https://openui5.hana.ondemand.com/resources/sap-ui-core.js" data-sap-ui-theme="sap_bluecrystal" data-sap-ui-libs="sap.m" data-sap-ui-xx-supportedLanguages="en" data-sap-ui-resourceroots='{"ToDoMVC": "./js"}'></script>
```
Comment (samccone) [inline examples/openui5/index.html]: can we just vendor this file ^

Comment (agraebe) [inline examples/openui5/index.html]: Unfortunately, that doesn't work. This file initiates lazy loading of many further files, depending on what components the app actually uses. The only way to have them files in there is to put the whole library of JS files in the commit.. That blows up the size though (as I figured after the initial PR).

Comment (samccone) [inline examples/openui5/index.html]: vendoring all the files is 100% expected :)

Don't worry (:

Comment (agraebe) [inline examples/openui5/index.html]: :+1: extracting the required files from the runtime ZIP and running tests right now.

**Comment by agraebe** [comment]
@samccone what do you think? Only the test left that's not passing?

**Comment by samccone** [comment]
yep, need to take it for a manual test :)

Thanks for all your work on this so far

**Comment by agraebe** [comment]
:+1: thanks for your support! let me know if you can replicate the error locally.. I'll try myself in the meantime as well. I want them all to pass on the CI landscpae as well ;)

**Comment by agraebe** [comment]
@samccone any proceedings from your side?

**Comment by agraebe** [comment]
@samccone :hourglass_flowing_sand: :walking:

**Comment by samccone** [comment]
yes I am going to take a look soon

sorry, if you would like to help feel free to look at any of the open PRs thanks!

**Comment by agraebe** [comment]
sure thing! I'll see if I can find something related :)

**Comment by agraebe** [comment]
@samccone just a friendly reminder.. would love to get this over with :)

**Comment by agraebe** [comment]
@samccone would you please respond? thanks..

**Comment by samccone** [comment]
hi @agraebe this is on my queue

**Comment by agraebe** [comment]
@samccone it's been now over a month. Could you please help to get this finally over with? Thanks..

File: examples/openui5/js/control/ToDoControl.js [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+			});
+
+			this.rerender();
+		},
+
+		_getAllToDosWithStatus: function (status) {
+			var elems = this.getToDosObject();
+			var toDoIds = [];
+
+			for (var i = 0; i < elems.length; i++) {
```
Comment (samccone) [inline examples/openui5/js/control/ToDoControl.js]: elms.forEach?

File: examples/openui5/js/control/ToDoControl.js [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+			this.rerender();
+		},
+
+		onToggleAllToDos: function () {
+			var toDoIds = this._getAllToDosWithStatus(false);
+			var elems = this.getToDosObject();
+
+			if (toDoIds.length > 0) {
+				// complete all uncompleted toDos
+				for (var i = 0; i < toDoIds.length; i++) {
```
Comment (samccone) [inline examples/openui5/js/control/ToDoControl.js]: toDoIds.forEach?

File: examples/openui5/js/control/ToDoControl.js [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
+					// fire changed event
+					this.fireEvent('completedToDoPressed', {
+						toDoId: toDoIds[i],
+						completed: true
+					});
+				}
+			} else {
+				// uncomplete all toDos
+				for (var j = 0; j < elems.length; j++) {
```
Comment (samccone) [inline examples/openui5/js/control/ToDoControl.js]: forEach?

**Comment by agraebe** [comment]
@samccone I replaced all the for loops with forEach .. all tests pass locally, but now another error appears:

```
TypeError: webdriver.WebDriver.Logs is not a function
    at getLastCounts (/home/travis/build/tastejs/todomvc/tests/node_modules/drool/lib/index.js:27:11)
    at getCounts (/home/travis/build/tastejs/todomvc/tests/node_modules/drool/lib/index.js:14:10)
    at Object.flow (/home/travis/build/tastejs/todomvc/tests/node_modules/drool/lib/index.js:47:3)
    at /home/travis/build/tastejs/todomvc/tests/memory.js:45:8
    at Array.forEach (native)
    at Object.<anonymous> (/home/travis/build/tastejs/todomvc/tests/memory.js:44:6)
    at Module._compile (module.js:413:34)
    at Object.Module._extensions..js (module.js:422:10)
    at Module.load (module.js:357:32)
    at Function.Module._load (module.js:314:12)
```

**Comment by addyosmani** [comment]
@agraebe I think the issue may have been addressed by https://github.com/tastejs/todomvc/pull/1606. Would you still be interested in updating your PR to resolve the conflicts against master?

File: index.html [hunk 7]
```diff
@@ -126,6 +126,9 @@
 								<li class="routing">
 									<a href="examples/troopjs_require/" data-source="https://github.com/troopjs/" data-content="TroopJS attempts to package popular front-end technologies and bind them with minimal effort for the developer. It includes jQuery for DOM manipulation, When.js for promises, RequireJS for modularity and Has.js for feature detection. On top, it includes Pub/Sub support, templating, weaving (widgets to DOM) and auto-wiring.">TroopJS + RequireJS</a>
 								</li>
+								<li class="routing">
+									<a href="examples/openui5/" data-source="https://github.com/sap/openui5" data-content="OpenUI5 lets you build enterprise-ready web applications, responsive to all devices, running on almost any browser of your choice.">OpenUI5</a>
+								</li>
```
Comment (arthurvr) [inline index.html]: Don't directly put the app here, all apps start in labs first :) We can discuss putting it in the "main" section after it's been test running online for a while and when we we're sure it's gonna stay maintained.

**Comment by arthurvr** [comment]
Hey, sorry this PR has been taking so long. Are you still interested in getting it updated? Thanks!

File: examples/openui5/js/app.js [hunk 8]
```diff
@@ -0,0 +1,43 @@
+/* global sap */
+
+(function () {
+	'use strict';
+
+	// Polyfill for .bind()
```
Comment (arthurvr) [inline examples/openui5/js/app.js]: Hm, is this really needed? We only care about supporting modern browsers :)

File: examples/openui5/js/control/ToDoControl.js [hunk 9]
```diff
@@ -0,0 +1,348 @@
...
+						title: {
+							type: 'string'
+						}
+					}
+				}
+			}
+		},
+
+		onAddToDo: function (evt) {
+			if (+evt.keyCode === ENTER_KEY) {
```
Comment (arthurvr) [inline examples/openui5/js/control/ToDoControl.js]: Hm, is `+` needed here? Is there any reason `keyCode` wouldn't be a number?

Comment (arthurvr) [inline examples/openui5/js/control/ToDoControl.js]: (same on line 89 with `title.length`)

Comment (arthurvr) [inline examples/openui5/js/control/ToDoControl.js]: And a bunch of other places :D

File: examples/openui5/js/control/ToDoControl.js [hunk 10]
```diff
@@ -0,0 +1,348 @@
...
+		 * Define how the ToDoMVC control should be rendered
+		 */
+		renderer: {
+			controlID,
+
+			render: function (oRm, oControl) {
+				controlID = oControl.getId();
+				toDos = oControl.getToDosObject();
+				filter = oControl.getFilter();
+				toDosAvailable = (toDos && toDos.length > 0) ? true : false;
```
Comment (arthurvr) [inline examples/openui5/js/control/ToDoControl.js]: I'm curious, why the `? true : false` here? `toDos && toDos.length > 0` is a boolean anyhow right?

**Pull request closed without merging**